[SPARK-4879] Use the Spark driver to authorize Hadoop commits.

mccheah opened the pull request:
This is a version of https://github.com/apache/spark/pull/4066/ which is up to date with master and has unit tests.

Previously, SparkHadoopWriter always committed its tasks without question. The problem is that when speculation is turned on, sometimes this can result in multiple tasks committing their output to the same
partition. Even though an HDFS-writing task may be re-launched due to speculation, the original task is not killed and may eventually commit as well.

This can cause strange race conditions where multiple tasks that commit interfere with each other, with the result being that some partition files are actually lost entirely. For more context on these kinds of scenarios, see the aforementioned JIRA ticket.

In Hadoop MapReduce jobs, the application master is a central coordinator that authorizes whether or not any given task can commit. Before a task commits its output, it queries the application master as to whether or not such a commit is safe, and the application master does bookkeeping as tasks are requesting commits. Duplicate tasks that would write to files that were already written to from other tasks are prohibited from committing.

This patch emulates that functionality - the crucial missing component was a central arbitrator, which is now a module called the OutputCommitCoordinator. The coordinator lives on the driver and the executors can obtain a reference to this actor and request its permission to commit. As tasks commit and are reported as completed successfully or unsuccessfully by the DAGScheduler, the commit coordinator is informed of the task completion events as well to update its internal state.

Future work includes more rigorous unit testing and extra optimizations should this patch cause a performance regression. It is unclear what the overall cost of communicating back to the driver on every hadoop-committing task will be. It's also important for those hitting this issue to backport this onto previous version of Spark because the bug has serious consequences, that is, data is lost.

cc @MingyuKim @JoshRosen  @ash211

mccheah committed:
[SPARK-4879] Use the Spark driver to authorize Hadoop commits.

Previously, SparkHadoopWriter always committed its tasks without
question. The problem is that when speculation is turned on, sometimes
this can result in multiple tasks committing their output to the same
partition. Even though an HDFS-writing task may be re-launched due to
speculation, the original task is not killed and may eventually commit
as well.

This can cause strange race conditions where multiple tasks that commit
interfere with each other, with the result being that some partition
files are actually lost entirely. For more context on these kinds of
scenarios, see the aforementioned JIRA ticket.

In Hadoop MapReduce jobs, the application master is a central
coordinator that authorizes whether or not any given task can commit.
Before a task commits its output, it queries the application master as
to whether or not such a commit is safe, and the application master does
bookeeping as tasks are requesting commits. Duplicate tasks that would
write to files that were already written to from other tasks are
prohibited from committing.

This patch emulates that functionality - the crucial missing component was
a central arbitrator, which is now a module called the OutputCommitCoordinator.
The coordinator lives on the driver and the executors can obtain a reference
to this actor and request its permission to commit. As tasks commit and are
reported as completed successfully or unsuccessfully by the DAGScheduler,
the commit coordinator is informed of the task completion events as well
to update its internal state.

mccheah committed:
Unit tests for OutputCommitCoordinator

mccheah committed:
Scalastyle fixes

mccheah committed:
Removing redundant accumulator in unit test

SparkQA commented:
[Test build #25940 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25940/consoleFull) for   PR 4155 at commit [`6b543ba`](https://github.com/apache/spark/commit/6b543bae5d9a6ea27408a57663986d3947ea7ab1).
 * This patch merges cleanly.

mccheah committed:
Removing whitespace modifications

mccheah committed:
Renaming oudated names for test function classes

SparkQA commented:
[Test build #25941 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25941/consoleFull) for   PR 4155 at commit [`1c2b219`](https://github.com/apache/spark/commit/1c2b21908225a644f268adb8e1e67a4479492532).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #25940 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25940/consoleFull)**     for PR 4155 at commit [`6b543ba`](https://github.com/apache/spark/commit/6b543bae5d9a6ea27408a57663986d3947ea7ab1)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25940/
Test FAILed.

SparkQA commented:
**[Test build #25941 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25941/consoleFull)**     for PR 4155 at commit [`1c2b219`](https://github.com/apache/spark/commit/1c2b21908225a644f268adb8e1e67a4479492532)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25941/
Test FAILed.

mccheah committed:
Moving the output commit coordinator from class into method.

The SparkHadoopWriter wasn't serializable because the commit coordinator
was a class member. It doesn't need to be, so just get it on demand when
committing the task.

SparkQA commented:
[Test build #25971 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25971/consoleFull) for   PR 4155 at commit [`f135a8e`](https://github.com/apache/spark/commit/f135a8ec6744e555a5d5d1ee1fdc71acf8d451d7).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #25971 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25971/consoleFull)**     for PR 4155 at commit [`f135a8e`](https://github.com/apache/spark/commit/f135a8ec6744e555a5d5d1ee1fdc71acf8d451d7)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25971/
Test FAILed.

mccheah committed:
TaskInfo can't be null in DAGSchedulerSuite

But we don't care about what's actually in the TaskInfo object - testing
OutputCommitCoordinator is done in OutputCommitCoordinatorSuite which is
the only code path that relies on the TaskInfo's fields. Nevertheless we
need the Task Info to not be null to allow that code path to not throw
an NPE.

SparkQA commented:
[Test build #25972 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25972/consoleFull) for   PR 4155 at commit [`abc7db4`](https://github.com/apache/spark/commit/abc7db44c145e10d6a26d71c63c3f541bab6d274).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #25972 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25972/consoleFull)**     for PR 4155 at commit [`abc7db4`](https://github.com/apache/spark/commit/abc7db44c145e10d6a26d71c63c3f541bab6d274)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25972/
Test FAILed.

mccheah commented:
Looks like the tests timed out. This change is probably a large performance bottleneck, as communication back to the driver on every commit task is expensive?

mccheah committed:
Making the OutputCommitCoordinatorMessage serializable

SparkQA commented:
[Test build #25983 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25983/consoleFull) for   PR 4155 at commit [`83de900`](https://github.com/apache/spark/commit/83de900eb60125e1190670e29938fc0b8634bdeb).
 * This patch merges cleanly.

mccheah commented:
Actually it just looks like one test is hanging, so likely something not being shut down properly.

mccheah committed:
Better OutputCommitCoordinatorActor stopping; simpler canCommit

Before, a poison pill message was sent to the actor. That is not the
paradigm that is used on other actors in Spark though, so making this
more like the e.g. MapOutputTracker Actor.

SparkQA commented:
[Test build #25985 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25985/consoleFull) for   PR 4155 at commit [`78eb1b5`](https://github.com/apache/spark/commit/78eb1b58cee0b5baf532ffde2a90e0fc31657d69).
 * This patch merges cleanly.

mccheah committed:
More OutputCommitCoordinator cleanup on stop()

SparkQA commented:
[Test build #25986 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25986/consoleFull) for   PR 4155 at commit [`9c6a4fa`](https://github.com/apache/spark/commit/9c6a4fa7be88428a0de0825d0c77bd990c6a1f2f).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #25983 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25983/consoleFull)**     for PR 4155 at commit [`83de900`](https://github.com/apache/spark/commit/83de900eb60125e1190670e29938fc0b8634bdeb)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25983/
Test FAILed.

mccheah committed:
Was mistakenly serializing the accumulator in test suite.

SparkQA commented:
[Test build #25991 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25991/consoleFull) for   PR 4155 at commit [`8d5a091`](https://github.com/apache/spark/commit/8d5a091bcb88b4376bffb0e16da9e388e7e79afe).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #25985 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25985/consoleFull)**     for PR 4155 at commit [`78eb1b5`](https://github.com/apache/spark/commit/78eb1b58cee0b5baf532ffde2a90e0fc31657d69)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25985/
Test FAILed.

SparkQA commented:
**[Test build #25986 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25986/consoleFull)**     for PR 4155 at commit [`9c6a4fa`](https://github.com/apache/spark/commit/9c6a4fa7be88428a0de0825d0c77bd990c6a1f2f)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25986/
Test FAILed.

SparkQA commented:
**[Test build #25991 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25991/consoleFull)**     for PR 4155 at commit [`8d5a091`](https://github.com/apache/spark/commit/8d5a091bcb88b4376bffb0e16da9e388e7e79afe)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25991/
Test FAILed.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,172 @@
+    stage: Int,
+    task: Long,
+    taskAttempt: Long)
+    extends OutputCommitCoordinationMessage
```
Minor style thing, but the `extends` should only be indented two spaces in order to visually distinguish it from the previous line.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,172 @@
+
+  // Initialized by SparkEnv
+  var coordinatorActor: ActorRef = _
+  val timeout = AkkaUtils.askTimeout(conf)
```
timeout, maxAttempts, and retryInterval should all be private.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,172 @@
+  private type TaskAttemptId = Long
+
+  private val authorizedCommittersByStage:
+  mutable.Map[StageId, mutable.Map[TaskId, TaskAttemptId]] = mutable.HashMap()
```
This line should be indented more, I think.  I don't know that we have a great style guideline for how to wrap really long explicit types like this.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,172 @@
+
+  def stageStart(stage: StageId) {
+    coordinatorActor ! StageStarted(stage)
+  }
```
Extra blank line after here.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,172 @@
+  private val authorizedCommittersByStage:
+  mutable.Map[StageId, mutable.Map[TaskId, TaskAttemptId]] = mutable.HashMap()
+
+  def stageStart(stage: StageId) {
```
For consistency, I'd do the `: Unit = {` style for these methods, too.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,172 @@
+      stage: StageId,
+      task: TaskId,
+      attempt: TaskAttemptId):
+      Boolean = {
```
Can move this return type to the previous line.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,172 @@
+
+private[spark] object OutputCommitCoordinator {
+
+  class OutputCommitCoordinatorActor(outputCommitCoordinator: OutputCommitCoordinator)
```
What's the motivation for defining this class inside of the `OutputCommitCoordinator` and having a `createActor` method instead of a regular constructor call?

mccheah replied:
I made all of the handleX methods on OutputCommitCoordinator private, so that the protocol enforced is to call the methods that send the message to the actor. The actor could just be built using a constructor, I agree - some misunderstanding on my end on how Scala works in SparkEnv.registerOrLookup, and will fix that.

However since @vanzin described an alternative approach to using the actor, this may change soon. Will be working on this today.

JoshRosen commented:
I'd be happy to help debug the test hang issue.  To confirm, though, have you tested this with your original bug reproduction from JIRA to see whether it resolves that issue?  Just want to check to see that we've actually solved the core problem here.

I tried logging into Jenkins to grab the logs from the test that timed out, but it looks like our log archiver doesn't work when tests time out.  Therefore, I guess I'd try running just that one problematic suite directly using sbt to see if you can reproduce the hang locally.

JoshRosen commented on core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:
```diff
@@ -0,0 +1,176 @@
+ * messages back to the DAG scheduler.
+ */
+class OutputCommitCoordinatorSuite
+    extends FunSuiteLike
```
Why FunSuiteLike instead of FunSuite?

JoshRosen commented on core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:
```diff
@@ -0,0 +1,176 @@
+  private class FailFirstTimeCommittingFunction
+      extends OutputCommittingFunction {
+    override def apply(ctxt: TaskContext, it: Iterator[Int]): Int = {
+      if (ctxt.attemptNumber == 0) {
```
This is a nice use of the new `attemptNumber` method (which is exactly why I merged that patch).  This sort of test is exactly why I'd also like to backport that patch to maintenance branches so that we can backport tests like this, too.

JoshRosen commented on core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:
```diff
@@ -0,0 +1,176 @@
+
+      private def execTasks(taskSet: TaskSet, attemptNumber: Int) {
+        var taskIndex = 0
+        taskSet.tasks.foreach(t => {
```
Minor style thing, but you can just use a curly brace w/ foreach and remove the second level of braces, e.g. `foreach{ t => `

mccheah commented:
Lots of comments to address, thanks for the detailed feedback @JoshRosen! One hanging cause is that the actor can be stopped multiple times in some of the YARN tests, so I'm patching that. Will push and see if that fixes Jenkins, and then address your comments accordingly.

We have yet to attempt to see if this fixes the actual issue on our side. I'm looking to pass jenkins before going for that test. But I do agree we need to verify against the core issue live before considering this PR a go.

mccheah commented:
Pushed to try to make Jenkins pass. If Jenkins passes I'll handle the comments so far.

mccheah committed:
Properly handling messages that could be sent after actor shutdown.

SparkQA commented:
[Test build #26006 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26006/consoleFull) for   PR 4155 at commit [`c334255`](https://github.com/apache/spark/commit/c3342552e03d690ac4beea939b5abd13363698c4).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26006 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26006/consoleFull) for   PR 4155 at commit [`c334255`](https://github.com/apache/spark/commit/c3342552e03d690ac4beea939b5abd13363698c4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class OutputCommitCoordinatorActor(outputCommitCoordinator: OutputCommitCoordinator)`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26006/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #26027 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26027/consoleFull) for   PR 4155 at commit [`c334255`](https://github.com/apache/spark/commit/c3342552e03d690ac4beea939b5abd13363698c4).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26027 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26027/consoleFull) for   PR 4155 at commit [`c334255`](https://github.com/apache/spark/commit/c3342552e03d690ac4beea939b5abd13363698c4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class OutputCommitCoordinatorActor(outputCommitCoordinator: OutputCommitCoordinator)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26027/
Test PASSed.

JoshRosen commented:
Woohoo, looks like this is passing tests!  The earlier failure was due to a known flaky streaming test.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,178 @@
+import scala.collection.mutable
+import scala.concurrent.duration.FiniteDuration
+
+import akka.actor.{PoisonPill, ActorRef, Actor}
```
super nit: sort imports (here and elsewhere)

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,178 @@
+  def stageStart(stage: StageId) {
+    sendToActor(StageStarted(stage))
+  }
+  def stageEnd(stage: StageId) {
```
super nit: missing an empty line between methods.

vanzin commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -19,12 +19,13 @@ package org.apache.spark.scheduler
 import org.scalatest.concurrent.Timeouts
 import org.scalatest.time.SpanSugar._
 
+import org.mockito.Mockito.mock
```
super nit: group with `org.scalatest`

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,178 @@
+    sendToActor(TaskCompleted(stage, task, attempt, successful))
+  }
+
+  def stop() {
```
Minor, but I think it's slightly weird that this class mixes methods that should only be called from the driver (such as `stop`) and methods that executors can call safely. Perhaps a check here that this is only being called on the driver side?

mccheah replied:
I'm not very experienced with Akka, but does stop() not need to be invoked on the actor on both the driver side and the executor side?

Granted the asking-permission-to-commit-output should only be done from executors while the other methods should only be called from the driver, and those conditions should be checked.

vanzin commented:
I had this (unbased) notion that tasks knew whether they were speculative or not, and thus the "non-speculative" ones would be able to avoid this extra hop to the driver and just commit things. But it seems that's not the case (and it sort of makes sense, in case the "speculative" task finishes first), so I guess this approach is fine.

One thing that worries me a bit is that I've been told before that akka actors' `onReceive` methods are single-threaded (meaning they'll never be called concurrently, even for messages coming from different remote endpoints). That can become a bottleneck on really large jobs. If that's really true, we should probably look at decoupling the processing of the message from the `onReceive` method so that multiple executors can be serviced concurrently.

mccheah commented:
I'm also concerned about the performance ramifications of this. We need to run performance benchmarks. However, the only critical path that is affected by this are tasks that are explicitly saving to Hadoop file. When a task completes, the DAGScheduler sends a message to the OutputCommitCoordinator actor so the DAGScheduler is not blocked by this logic.

We do actually need the processing to be single threaded, as trying to coordinate synchronization on the centralized arbitration logic is a bit of a nightmare. I mean, we could allow multiple threads to access the internal state of OutputCommitCoordinator and implement appropriate synchronization logic. I considered an optimization where the driver broadcasts to executors when tasks are being speculated, and the executors of the original tasks would know to check the commit authorization, and skip it for tasks that don't have speculated copies. There's a lot of race conditions that arise from that though, which further underlines the need to centralize everything.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -808,6 +810,7 @@ class DAGScheduler(
     // will be posted, which should always come after a corresponding SparkListenerStageSubmitted
     // event.
     stage.latestInfo = StageInfo.fromStage(stage, Some(partitionsToCompute.size))
+    outputCommitCoordinator.stageStart(stage.id)
```
I wonder if it wouldn't be better to use a `SparkListener` to reduce coupling. Although that would potentially introduce race conditions in the code (since `LiveListenerBus` fires events on a separate thread).

lianhuiwang replied:
yes, i agree with @vanzin 's Opinion.

JoshRosen replied:
I think that this introduces a race between commit requests and the stage start event.  If the listener bus is slow in delivering events, then it's possible that the output commit coordinator could receive a commit request via Akka for a stage that it doesn't know about yet.

vanzin commented on core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala:
```diff
@@ -106,18 +107,25 @@ class SparkHadoopWriter(@transient jobConf: JobConf)
+      val outputCommitCoordinator = SparkEnv.get.outputCommitCoordinator
+      val conf = SparkEnv.get.conf
+      val canCommit: Boolean = outputCommitCoordinator.canCommit(jobID, splitID, attemptID)
+      if (canCommit) {
```
Hmm. I wonder if this can be a problem. Given the following timeline:

    
    Time ->

    (1)--------(2)--------(3)
    
    (4)--------------(5)
    
1: task 1 start
2. task 1 asks for permission to commit, it's granted
3. task 1 fails to commit
4. task 2 starts (doing same work as task 1)
5. task 2 asks for permission to commit, it's denied

Wouldn't this code force a new task to be run to recompute everything? Also, wouldn't task 2 actually report itself as successful, and break things, since there is a successful task for that particular split, but it was never committed?

mccheah replied:
It would force a new task to recompute everything, but this does highlight that task 2 should throw an error, @JoshRosen?

mccheah replied:
But another follow-up question - how did we get away with this case before? If committer.needsTaskCommit returns false in SparkHadoopWriter it also does not throw an error. But if some other task that had committer.needsTaskCommit returned true but failed to commit the output, are we in the same situation as what @vanzin described albeit in a slightly different code path?

Or to be more explicit, take the 5 steps but slightly adjust them and assume the code I added in this patch doesn't exist:

1: Task 1 start
2. Task 1 asks committer.needsTaskCommit() which returns true
3. task 1 fails to commit
4. Task 2 starts
5. Task 2 asks committer.needsTaskCommit but that returns false

vanzin replied:
The original `committer.needsTaskCommit` is a Hadoop API method, so it follows the Hadoop semantics. The API docs for the method are not very helpful, so hopefully somebody knows what they are.

mccheah replied:
@JoshRosen, @vanzin's concern still holds. I'm going to change this so it throws an exception on the commit denial.

JoshRosen replied:
Ah, I missed this discussion since it got collapsed under a changed diff.

I'll have to revisit the docs / usage to confirm this, but I think that the Hadoop `needsTaskCommit` was a necessary but not sufficient condition for committing where `needsTaskCommit` should always return `true` for a task that has not had its output committed (in fact, I think it would be valid for it to always return `true`, even for already-committed output).

I agree that task 2 should throw an error, since it seems like doing otherwise would also lead to missing output: as soon as the scheduler sees one successful completion for a task, it won't re-run that task, so we need it to be the case that "successful task completion" implies "output committed."

To run with your example, I think that task 2 would see `needsTaskCommit = false` because the output hasn't been committed yet, but there's still a problem that can lead to missing output.  Imagine that we had this interleaving:

1: Tasks 1 and 2 start
2. Task 1 asks committer.needsTaskCommit() which returns true
3. Task 2 asks committer.needsTaskCommit(), which returns false.  It exits without throwing an exception and reports success.
4. Task 1 fails to commit

In this case, one copy of the task has reported success even though output was not committed, so that partition will be missing because we won't re-schedule an attempt to commit.

So, I agree with @vanzin: if the DAGScheduler did not authorize the commit, then we should throw an exception.  I think that this exception will hopefully be rare in practice because needsTaskCommit should ideally return `false` for tasks that are definitely committed.

vanzin replied:
Man, github makes it *really* hard to find this discussion. :-/

Anyway, just one comment:
> if the DAGScheduler did not authorize the commit, then we should throw an exception

As long as that doesn't cause the driver's "task error count" go up, then fine. Otherwise, we probably need some new state (e.g. "task gave up") that doesn't count as an error.

JoshRosen replied:
> Otherwise, we probably need some new state (e.g. "task gave up") that doesn't count as an error.

We have the `TaskEndReason` mechanism for special-handling of certain kinds of task failures, so maybe we can add a new reason and update the corresponding [handlers in TaskSetManager](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L605) and [DAGScheduler](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L941).

We might still end up having to throw an exception, but we can throw a more specific one that we can catch and transform into the right TaskEndReason.

JoshRosen replied:
For instance, there's code in Executor.scala to catch certain exceptions and translate them into TaskEndReasons: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/Executor.scala#L243

mccheah replied:
Makes sense.

I'm trying to implement this change now but am getting tripped up in unit tests again. Now, the unit tests need to capture the logic all the way down to the Executor. It looks like this workflow requires something closer to an end-to-end test now. Are there any unit tests that test scenarios with speculation? How should my existing OutputCommitCoordinator suite be adjusted?

vanzin commented:
> We do actually need the processing to be single threaded, as trying to coordinate synchronization on the centralized arbitration logic is a bit of a nightmare.

I'm not so convinced; you'd only have a conflict if two tasks are concurrently asking to update the state of the same split ID. Otherwise, state updates can happen in parallel.

e.g. if you know all the split IDs up front, you can initialize the data structure to hold all the state; when a commit request arrives, you only lock that particular state object. So requests that arrive for other split IDs can be processed in parallel.

(If you don't know all the split IDs up front, you can use something simple like `ConcurrentHashMap` or `ConcurrentSkipListMap` depending on what performance characteristics you want.)

rxin commented on core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:
```diff
@@ -0,0 +1,176 @@
+  test("If commit fails, if task is retried it should not be locked, and will succeed.") {
+    val rdd = sc.parallelize(Seq(1), 1)
+    sc.runJob(rdd, new FailFirstTimeCommittingFunction)
+    assert(accum.value == 1)
```
=== here

rxin commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -208,7 +196,7 @@ class DAGSchedulerSuite extends FunSuiteLike  with BeforeAndAfter with LocalSpar
     for ((result, i) <- results.zipWithIndex) {
       if (i < taskSet.tasks.size) {
-        runEvent(CompletionEvent(taskSet.tasks(i), result._1, result._2, null, null, null))
+        runEvent(CompletionEvent(taskSet.tasks(i), result._1, result._2, null, createFakeTaskInfo, null))
```
you want createFakeTaskInfo(), not createFakeTaskInfo

rxin commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,178 @@
+        sender ! true
+    }
+  }
+  def createActor(coordinator: OutputCommitCoordinator): OutputCommitCoordinatorActor = {
```
add a blank line

mccheah commented:
@vanzin I can attempt to make OutputCommitCoordinator more multithreaded as you suggest. Do we have an example somewhere of Spark executors calling back to the driver where the Driver handles those kinds of messages in a multithreaded way?

That might be a confusing question, so another way to put it: can we achieve this communication without remote actors? From what I understand Spark does all of its remote communication through remote actors.

mccheah committed:
Using more concurrency to process OutputCommitCoordinator requests.

The OutputCommitCoordinator now delegates commit authorization to a
thread pool. The synchronization at least allows multiple tasks that are
committing to different partitions to have their requests processed
concurrently. There is still a single message queue in the actor but the
message processing should be fast if it is just submitting a Runnable.

Also some style fixes in here.

SparkQA commented:
[Test build #26075 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26075/consoleFull) for   PR 4155 at commit [`d431144`](https://github.com/apache/spark/commit/d4311443b1aa652f103c46f6baae2f0d9721564f).
 * This patch merges cleanly.

mccheah commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,252 @@
+    coordinatorActor = Some(actor)
+    executorRequestHandlingThreadPool = {
+      if (isDriver) {
+        Some(Executors.newFixedThreadPool(4))
```
How should the number of threads be configured? Not submitting until we address this, I just chose an arbitrary number for now.

JoshRosen replied:
Do you think that the thread-pool here is going to buy us a performance benefit, or will we experience a bottleneck in the single-threaded message dispatch that feeds into the thread pool?

Also, if we do decide to use thread pools, we should use the Utils function that makes them create daemon threads: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/util/Utils.scala#L784

mccheah replied:
I would be surprised if submitting the job took longer than the job itself. The time spent to compute the canCommit computation should be at least the same amount of time to submit the job. Given this I think the parallelism will help. @vanzin what do you think?

vanzin replied:
While it's hard to tell without some code to prove it, I'd be surprised if the executor dispatch were slower than the akka pipeline here. In general, where possible, I think it's a good idea to decouple receiving of messages from processing them, just to avoid running into scalability issues with large jobs.

As long as that doesn't result in the creating of many purpose-specific thread pools all over the code base, though. :-)

SparkQA commented:
[Test build #26075 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26075/consoleFull) for   PR 4155 at commit [`d431144`](https://github.com/apache/spark/commit/d4311443b1aa652f103c46f6baae2f0d9721564f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class AskCommitRunnable(`
  * `  class OutputCommitCoordinatorActor(outputCommitCoordinator: OutputCommitCoordinator)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26075/
Test PASSed.

vanzin commented:
> Do we have an example somewhere of Spark executors calling back to the driver where the Driver handles those kinds of messages in a multithreaded way?

I don't remember any off the top of my head; in general, all the actors I've seen just handle things inline, for better or worse. Haven't looked at your latest update, but the usual way would be to have an executor service and submit tasks to it as akka messages are received. Not sure about the akka semantics (e.g. how to reply to the sender) from the worker thread, though.

mccheah commented:
@vanzin that's pretty much what I went with. The actor will receive the message and for commit permission requests they're farmed off to a thread pool.

mccheah commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,252 @@
+  def stop(): Unit = {
+    executorRequestHandlingThreadPool.foreach { pool =>
+      pool.shutdownNow()
+      pool.awaitTermination(10, TimeUnit.SECONDS)
```
What timeout should be used here?

mccheah committed:
Throwing exception if SparkHadoopWriter commit denied

However, the Executor will not treat this like any other
ExceptionFailure, but rather send a specific TaskEndReason to the
driver. The driver simply ignores the task that failed to commit after
logging. If the task that attempted to commit first fails, it will be
resubmitted.

TODO unit tests, the current unit tests don't capture this workflow all
the way down to the Executor level.

mccheah commented:
This is a work in progress. In particular, the OutputCommitCoordinatorSuite isn't quite testing the right thing now. I don't exactly know how to test the full stack (as now this requires actual executors to throw the TaskEndReason back to the driver) while having the scheduler somehow submit two copies of a task in speculation-like fashion. Any suggestions here?

SparkQA commented:
[Test build #26173 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26173/consoleFull) for   PR 4155 at commit [`1df2a91`](https://github.com/apache/spark/commit/1df2a91eb39300a32ad095b37a04846d135e2cc5).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26173 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26173/consoleFull) for   PR 4155 at commit [`1df2a91`](https://github.com/apache/spark/commit/1df2a91eb39300a32ad095b37a04846d135e2cc5).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CommitDeniedException(msg: String, jobID: Int, splitID: Int, attemptID: Int)`
  * `case class TaskCommitDenied(`
  * `  class AskCommitRunnable(`
  * `  class OutputCommitCoordinatorActor(outputCommitCoordinator: OutputCommitCoordinator)`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26173/
Test FAILed.

mccheah committed:
Fixing scalastyle

mccheah commented on core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:
```diff
@@ -0,0 +1,177 @@
+ * Unit tests for the output commit coordination functionality. Overrides the
+ * SchedulerImpl to just run the tasks directly and send completion or error
+ * messages back to the DAG scheduler.
+ */
```
So this is no longer testing the right thing. But I haven't been able to find an example of a unit test that overrides some of the SchedulerImpl's methods but keeps everything else the same as the default setup. Any suggestions?

SparkQA commented:
[Test build #26176 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26176/consoleFull) for   PR 4155 at commit [`9fe6495`](https://github.com/apache/spark/commit/9fe64953aa437ed1ed88a294e04129afc8f2bbb5).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26176 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26176/consoleFull) for   PR 4155 at commit [`9fe6495`](https://github.com/apache/spark/commit/9fe64953aa437ed1ed88a294e04129afc8f2bbb5).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CommitDeniedException(msg: String, jobID: Int, splitID: Int, attemptID: Int)`
  * `case class TaskCommitDenied(`
  * `  class AskCommitRunnable(`
  * `  class OutputCommitCoordinatorActor(outputCommitCoordinator: OutputCommitCoordinator)`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26176/
Test FAILed.

mccheah committed:
Fixing compiler error

SparkQA commented:
[Test build #26179 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26179/consoleFull) for   PR 4155 at commit [`d63f63f`](https://github.com/apache/spark/commit/d63f63f5769a29d9377b15f3025726477226ca88).
 * This patch merges cleanly.

JoshRosen commented:
> I don't exactly know how to test the full stack (as now this requires actual executors to throw the TaskEndReason back to the driver) while having the scheduler somehow submit two copies of a task in speculation-like fashion. Any suggestions here?

This is a tricky problem, since I don't think we have any end-to-end tests of task speculation with non-mocked components because Spark won't schedule speculative tasks on the same host.  Maybe we could add a flag to disable this "same host exclusion" for tests.

There are some tricky testing issues here and I haven't fully thought through them yet.

aarondav commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,258 @@
+  // while still setting option itself to Some(...) or None.
+  private class LockableAttemptId(var value: Option[TaskAttemptId])
+
+  private type CommittersByStageHashMap =
```
This code seems somewhat convoluted for the sake of multithreaded performance. Was there an observed performance degradation due to this patch?

Generally, I would expect a simple java.util.HashMap being accessed by a single thread to be able to handle on the order of 100,000 to 1,000,000 operations per second, and our workload seems to be depending only on the number of tasks completing saveAs* operations. If this is a bottleneck, then the data being written is likely trivially small.

vanzin replied:
> Generally, I would expect a simple java.util.HashMap being accessed by a single thread to be able to handle on the order of 100,000 to 1,000,000

Hi @aarondav, I agree with this if the work being done single-threadedly here is just the state updates. But what about the rest of the akka stack? Does that influence things here? I'm not too familiar with akka internals, but if queuing / serializing the reply message back to sender is also done in a single thread, then this pattern might indeed be worth it.

But anyway, as I've said before, it's hard to know without numbers. It would probably not be hard to write a toy benchmark for it. It's also probably fine to push the simple, single-threaded version first and enhance it later if needed.

JoshRosen replied:
I would be very surprised if the thread pool resulted in a significant performance improvement.  The extra overheads of submitting tasks to the pool, working with immutable maps, etc. might actually harm performance unless we recover other time due to the parallelism.

I'm not super-familiar with Akka internals, but the fact that an actor processes messages serially doesn't necessarily mean that the lower-level message transport code needs to be single-threaded.

Until we know that this gives a performance improvement or we observe a performance regression due to this patch, I'd like to roll back these thread-safety and parallelism changes and stick with a simpler single-threaded implementation; I think it will be easier to understand and less bug-prone.

mccheah replied:
I'm rolling this back right now - I was trying to fix my unit tests and was confused by my own logic, which is indicative of something bad =)

SparkQA commented:
**[Test build #26179 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26179/consoleFull)**     for PR 4155 at commit [`d63f63f`](https://github.com/apache/spark/commit/d63f63f5769a29d9377b15f3025726477226ca88)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26179/
Test FAILed.

mccheah commented:
Jenkins, retest this please

SparkQA commented:
[Test build #26192 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26192/consoleFull) for   PR 4155 at commit [`d63f63f`](https://github.com/apache/spark/commit/d63f63f5769a29d9377b15f3025726477226ca88).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26193 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26193/consoleFull) for   PR 4155 at commit [`d63f63f`](https://github.com/apache/spark/commit/d63f63f5769a29d9377b15f3025726477226ca88).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26192 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26192/consoleFull) for   PR 4155 at commit [`d63f63f`](https://github.com/apache/spark/commit/d63f63f5769a29d9377b15f3025726477226ca88).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CommitDeniedException(msg: String, jobID: Int, splitID: Int, attemptID: Int)`
  * `case class TaskCommitDenied(`
  * `  class AskCommitRunnable(`
  * `  class OutputCommitCoordinatorActor(outputCommitCoordinator: OutputCommitCoordinator)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26192/
Test PASSed.

SparkQA commented:
[Test build #26193 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26193/consoleFull) for   PR 4155 at commit [`d63f63f`](https://github.com/apache/spark/commit/d63f63f5769a29d9377b15f3025726477226ca88).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CommitDeniedException(msg: String, jobID: Int, splitID: Int, attemptID: Int)`
  * `case class TaskCommitDenied(`
  * `  class AskCommitRunnable(`
  * `  class OutputCommitCoordinatorActor(outputCommitCoordinator: OutputCommitCoordinator)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26193/
Test PASSed.

sryza commented on core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala:
```diff
@@ -106,18 +107,30 @@ class SparkHadoopWriter(@transient jobConf: JobConf)
+          }
         }
+      } else {
+        val msg: String = s"$taID: Not committed because the driver did not authorize commit"
```
We should be calling abortTask in this case as well.

sryza commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -63,7 +63,7 @@ class DAGScheduler(
     blockManagerMaster: BlockManagerMaster,
     env: SparkEnv,
-    clock: Clock = SystemClock)
+    clock: org.apache.spark.util.Clock = SystemClock)
```
Is there another conflicting `Clock` imported?

JoshRosen replied:
Spark has a few different `Clock` classes, so maybe this was to disambiguate: https://issues.apache.org/jira/browse/SPARK-4682

mccheah replied:
Yeah this was weird, it wouldn't compile without me quantifying the class here. Although that could be IntelliJ organizing my imports in a non-ideal way.

sryza commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -908,6 +912,11 @@ class DAGScheduler(
     val taskType = Utils.getFormattedClassName(task)
+    val isSuccess = event.reason == Success
+
+    outputCommitCoordinator.taskCompleted(stageId, event.taskInfo.taskId,
```
There should only be two spaces of indentation on the following lines.  Also, it looks like some of the args can be pulled back up without hitting 100 characters.

sryza commented on core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala:
```diff
@@ -0,0 +1,258 @@
+  extends OutputCommitCoordinationMessage
+
+/**
+ * Authority that decides whether tasks can commit output to HDFS.
```
Can this comment be augmented a little bit to describe a situation in which committing might be denied?

squito commented on core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:
```diff
@@ -0,0 +1,177 @@
+    with LocalSparkContext
+    with Timeouts {
+
+  val conf = new SparkConf().set("spark.localExecution.enabled", "true")
```
this doesn't have any effect, does it?  If I'm following things correctly, your call to `runJob` has `allowLocal = false`, which makes this config irrelevant

squito commented on core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:
```diff
@@ -0,0 +1,177 @@
+          super.commitTask(taskAttemptContext)
+        }
+      }
+      runCommitWithProvidedCommitter(ctxt, it, outputCommitter)
```
why do you need the anonymous subclass of `FakeOutputCommitter`?  isn't this just
`runCommitWithProvidedCommitter(ctxt, it, new FakeOutputCommitter)`?

squito commented:
I worry about how complicated the test is, and how much it needs to muck around with internals ... it may be hard to keep up to date as those internals change.  And, I'm not sure it actually tests the behavior you are trying to verify.  I just grabbed only `OutputCommitCoordinatorSuite` & `DAGSchedulerSingleThreadedProcessLoop`, without any of the other changes -- and the tests pass.  (which is not to say the test is useless.)

Unfortunately, I don't have any good ideas for a better test.  I've been thinking about it for a while, and still haven't come up with anything.  You could run a speculative job which writes to hadoop and see if it works, though that could easily pass even if this didn't always work.  Maybe `OutputCommitCoordinator` needs some special hooks for testing.  I'll keep thinking about it, but wanted to get others thoughts.

mccheah commented:
Regarding the tests, I note that this is currently NOT testing the right thing. It was, until I added extra logic down in the Executors, which is now bypassed by the way I'm mocking things.

I'm going to try using Mockito.spy() to wrap the existing TaskScheduler so that all of it and its downstream components' configurations are preserved. Let's see how that turns out.

aarondav commented on core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala:
```diff
@@ -106,18 +107,30 @@ class SparkHadoopWriter(@transient jobConf: JobConf)
+      val msg: String = s"No need to commit output of task" +
+        " because needsTaskCommit=false: ${taID.value}"
+      logInfo(msg)
+      throw new CommitDeniedException(msg, jobID, splitID, attemptID)
```
I think if needsTaskCommit=false, we should maintain the old behavior of not throwing an exception.

JoshRosen replied:
My understanding is that if `needsTaskCommit` is false, then the task's output has been successfully committed and therefore we don't ever need to throw an exception from this branch.  However, we still need to throw an exception in the case where the commit is denied by the driver.

aarondav replied:
Agreed.

JoshRosen commented:
> The users on my side have been able to reproduce the missing files issue reliably, so we may just have to live with an empirical verification and be done with it.

Given that the actual bug is non-deterministic, I think we could be okay without a regression test that reliably reproduces this issue.  There might be some value in a non-deterministic regression test, though, as long as it detects the bug with sufficiently high probability, since we'd eventually catch any regression by noticing that the test had become flaky in Jenkins.

Unless we can come up with a better test, in the immediate term I'm okay with having unit tests for the individual components and an empirical verification using your reproduction.  Even though they aren't regression tests, the new tests added here will be helpful for preventing regressions if anyone changes the OutputCommitCoordinator logic.

mccheah committed:
Writing proper unit test for OutputCommitCoordinator and fixing bugs.

SparkQA commented:
[Test build #26347 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26347/consoleFull) for   PR 4155 at commit [`60a47f4`](https://github.com/apache/spark/commit/60a47f4cd71d1ada6750b35e1b9e16e37c1caab4).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26347 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26347/consoleFull) for   PR 4155 at commit [`60a47f4`](https://github.com/apache/spark/commit/60a47f4cd71d1ada6750b35e1b9e16e37c1caab4).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CommitDeniedException(msg: String, jobID: Int, splitID: Int, attemptID: Int)`
  * `case class TaskCommitDenied(`
  * `  class OutputCommitCoordinatorActor(outputCommitCoordinator: OutputCommitCoordinator)`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26347/
Test FAILed.

mccheah committed:
Fixing a scalastyle error

SparkQA commented:
[Test build #26348 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26348/consoleFull) for   PR 4155 at commit [`594e41a`](https://github.com/apache/spark/commit/594e41abecf5a48084608ab20112f884f28fc920).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #26348 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26348/consoleFull)**     for PR 4155 at commit [`594e41a`](https://github.com/apache/spark/commit/594e41abecf5a48084608ab20112f884f28fc920)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26348/
Test FAILed.

mccheah commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #26365 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26365/consoleFull) for   PR 4155 at commit [`594e41a`](https://github.com/apache/spark/commit/594e41abecf5a48084608ab20112f884f28fc920).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #26365 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26365/consoleFull)**     for PR 4155 at commit [`594e41a`](https://github.com/apache/spark/commit/594e41abecf5a48084608ab20112f884f28fc920)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26365/
Test FAILed.

JoshRosen commented:
Since we're nearing a working solution, I'd like to aim to include this patch in 1.3.  Since this patch involves major changes to the output commit code, I'd like to propose that we "feature-flag" the output committer behind a new configuration option.  I think (but please correct me if I'm wrong) that we can safely bypass the new driver coordination when speculation is disabled (this should also alleviate some of the performance impact concerns that have been raised here).  When speculation is enabled, we should perform the new checks by default, but should have an emergency "escape-hatch" option to bypass the new checks in case they present problems or contain bugs.  Therefore, I think the new setting's default value could be set to `spark.speculation.enabled`'s value.  I'm open to naming settings for the new configuration, but I was thinking that it could be something like `spark.hadoop.outputCommitCoordination.enabled` (maybe that's too verbose).  It's a toss-up on whether we'd want to include this in `configuration.md`, since I can't imagine that users would want to disable it unless we found bugs in the new code.

mccheah commented:
@JoshRosen I'll be out of the country for two weeks on a business trip. I'll attempt to push this through the hanging unit tests that are coming up now, but I can't guarantee I'll be able to finish it soon; if this is needed for 1.3.0, might I suggest you finish off the work from here? Sorry for the inconvenience.

JoshRosen commented:
@mccheah I'm fine with picking this up and doing the final work to finish it.  Just ping me once you've pushed your final changes and I'll pull these commits into my PR and continue work there.

mccheah commented:
Hm, the tests that seem to be running last, don't hang on my machine. Maybe I just got really unlucky?

mccheah commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #26428 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26428/consoleFull) for   PR 4155 at commit [`594e41a`](https://github.com/apache/spark/commit/594e41abecf5a48084608ab20112f884f28fc920).
 * This patch merges cleanly.

mccheah commented:
Tried running the Streaming CheckpointSuite locally, and it broke because of the new CommitDeniedException logic I added. Don't have any ideas as to how this happens except that streaming might not be using SparkHadoopWriter in a way that is compatible with this design, perhaps...

I don't think I'll be able to take this any further. Feel free to pick things up from here, @JoshRosen.

SparkQA commented:
**[Test build #26428 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26428/consoleFull)**     for PR 4155 at commit [`594e41a`](https://github.com/apache/spark/commit/594e41abecf5a48084608ab20112f884f28fc920)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26428/
Test FAILed.

JoshRosen commented:
I'm continuing work on this output commit coordination patch over at #4066, in case anyone would like to help with further review.

JoshRosen commented:
([Cross-posted from JIRA](https://issues.apache.org/jira/browse/SPARK-4879?focusedCommentId=14315077&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14315077))

This issue is really hard to reproduce, but I managed to trigger the original bug as part of the testing for my patch. Here's what I ran:

```bash
~/spark-1.3.0-SNAPSHOT-bin-1.0.4/bin/spark-shell --conf spark.speculation.multiplier=1 --conf spark.speculation.quantile=0.01 --conf spark.speculation=true --conf  spark.hadoop.outputCommitCoordination.enabled=false
```

```scala
val numTasks = 100
val numTrials = 100
val outputPath = "/output-committer-bug-"
val sleepDuration = 1000

for (trial <- 0 to (numTrials - 1)) {
  val outputLocation = outputPath + trial
  sc.parallelize(1 to numTasks, numTasks).mapPartitionsWithContext { case (ctx, iter) =>
    if (ctx.partitionId % 5 == 0) {
      if (ctx.attemptNumber == 0) {  // If this is the first attempt, run slow
       Thread.sleep(sleepDuration)
      }
    }
    iter
  }.map(identity).saveAsTextFile(outputLocation)
  Thread.sleep(sleepDuration * 2)
  println("TESTING OUTPUT OF TRIAL " + trial)
  val savedData = sc.textFile(outputLocation).map(_.toInt).collect()
  if (savedData.toSet != (1 to numTasks).toSet) {
    println("MISSING: " + ((1 to numTasks).toSet -- savedData.toSet))
    assert(false)
  }
  println("-" * 80)
}
```

It took 22 runs until I actually observed missing output partitions (several of the earlier runs threw spurious exceptions and didn't have missing outputs):

```
[...]
15/02/10 22:17:21 INFO scheduler.DAGScheduler: Job 66 finished: saveAsTextFile at <console>:39, took 2.479592 s
15/02/10 22:17:21 WARN scheduler.TaskSetManager: Lost task 75.0 in stage 66.0 (TID 6861, ip-172-31-1-124.us-west-2.compute.internal): java.io.IOException: Failed to save output of task: attempt_201502102217_0066_m_000075_6861
        at org.apache.hadoop.mapred.FileOutputCommitter.moveTaskOutputs(FileOutputCommitter.java:160)
        at org.apache.hadoop.mapred.FileOutputCommitter.moveTaskOutputs(FileOutputCommitter.java:172)
        at org.apache.hadoop.mapred.FileOutputCommitter.commitTask(FileOutputCommitter.java:132)
        at org.apache.spark.SparkHadoopWriter.performCommit$1(SparkHadoopWriter.scala:113)
        at org.apache.spark.SparkHadoopWriter.commit(SparkHadoopWriter.scala:150)
        at org.apache.spark.rdd.PairRDDFunctions$$anonfun$13.apply(PairRDDFunctions.scala:1082)
        at org.apache.spark.rdd.PairRDDFunctions$$anonfun$13.apply(PairRDDFunctions.scala:1059)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:61)
        at org.apache.spark.scheduler.Task.run(Task.scala:64)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:197)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)

15/02/10 22:17:21 WARN scheduler.TaskSetManager: Lost task 80.0 in stage 66.0 (TID 6866, ip-172-31-11-151.us-west-2.compute.internal): java.io.IOException: The temporary job-output directory hdfs://ec2-54-213-142-80.us-west-2.compute.amazonaws.com:9000/output-committer-bug-22/_temporary doesn't exist!
        at org.apache.hadoop.mapred.FileOutputCommitter.getWorkPath(FileOutputCommitter.java:250)
        at org.apache.hadoop.mapred.FileOutputFormat.getTaskOutputPath(FileOutputFormat.java:244)
        at org.apache.hadoop.mapred.TextOutputFormat.getRecordWriter(TextOutputFormat.java:116)
        at org.apache.spark.SparkHadoopWriter.open(SparkHadoopWriter.scala:91)
        at org.apache.spark.rdd.PairRDDFunctions$$anonfun$13.apply(PairRDDFunctions.scala:1068)
        at org.apache.spark.rdd.PairRDDFunctions$$anonfun$13.apply(PairRDDFunctions.scala:1059)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:61)
        at org.apache.spark.scheduler.Task.run(Task.scala:64)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:197)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)

15/02/10 22:17:21 INFO scheduler.TaskSetManager: Lost task 85.0 in stage 66.0 (TID 6871) on executor ip-172-31-1-124.us-west-2.compute.internal: java.io.IOException (The temporary job-output directory hdfs://ec2-54-213-142-80.us-west-2.compute.amazonaws.com:9000/output-committer-bug-22/_temporary doesn't exist!) [duplicate 1]
15/02/10 22:17:21 INFO scheduler.TaskSetManager: Lost task 90.0 in stage 66.0 (TID 6876) on executor ip-172-31-1-124.us-west-2.compute.internal: java.io.IOException (The temporary job-output directory hdfs://ec2-54-213-142-80.us-west-2.compute.amazonaws.com:9000/output-committer-bug-22/_temporary doesn't exist!) [duplicate 2]
15/02/10 22:17:21 INFO scheduler.TaskSetManager: Lost task 95.0 in stage 66.0 (TID 6881) on executor ip-172-31-11-151.us-west-2.compute.internal: java.io.IOException (The temporary job-output directory hdfs://ec2-54-213-142-80.us-west-2.compute.amazonaws.com:9000/output-committer-bug-22/_temporary doesn't exist!) [duplicate 3]
15/02/10 22:17:21 INFO scheduler.TaskSchedulerImpl: Removed TaskSet 66.0, whose tasks have all completed, from pool
TESTING OUTPUT OF TRIAL 22
[...]
15/02/10 22:17:23 INFO scheduler.TaskSchedulerImpl: Removed TaskSet 67.0, whose tasks have all completed, from pool
15/02/10 22:17:23 INFO scheduler.DAGScheduler: Stage 67 (collect at <console>:42) finished in 0.158 s
15/02/10 22:17:23 INFO scheduler.DAGScheduler: Job 67 finished: collect at <console>:42, took 0.165580 s
MISSING: Set(76)
java.lang.AssertionError: assertion failed
[...]
```

And I confirmed that it's missing in HDFS:

```
~/ephemeral-hdfs/bin/hadoop fs -ls /output-committer-bug-22 | grep part | wc -l
Warning: $HADOOP_HOME is deprecated.

99
```

To test my patch, I'm going to remove the flag that disables it and run this for a huge number of trials to ensure that we don't hit the missing output bug.

andrewor14 commented:
@mccheah looks like this is being superseded by #4066. Shall we close this PR then?
